### DISCUSSION TITLE
How to define features of each node in a heterogeneous graph?

### DISCUSSION THREAD

**Discussion by zzzLemon** [pyg-team/pytorch_geometric#6136] [Q&A] [answered]
Hi,

thanks at first for your great work on pyg. I am trying to generate heterogeneous graphs for GCN training following the tutorial [https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html](url).

I have some questions regarding the heterogeneous graph:
First, how to set features of each node individually? For example, "data['paper'].x = ... # [num_papers, num_features_paper]", if the node type "paper" has three features/attributes, let's say "name","year","pages", and if I have 2 "paper" nodes with different feature values, how can I set their features like {Node 0, "name":"1","year":2000,"page":1}, {Node 1, "name":"2","year":2001,"page":2}, and connect them through "data[Node 0, 'connect_to', Node 1].edge_index=..."?

Second, I only have one edge type, is it ok to set all edge as the same string e.g. as "connect_to"?

Thanks for your answer!

**Comment by rusty1s** [comment, marked as answer]
You would need to create a matrix out of your feature dictionary. If you have
```python
features = [{"node": 0, "name":"1","year":2000,"page":1}, ...]
```
then you can do
```python
xs = [None] * len(features)
for feature in features:
    xs[feature["node"]] = [int(feature["name"]), feature["year"], feature["page"]]
x = torch.tensor(xs)
```

> Second, I only have one edge type, is it ok to set all edge as the same string e.g. as "connect_to"?

Yes

**Reply by zzzLemon** [reply]
Hi,

thanks for your answer! I'm still wondering how the edge_index works for HeteroData. Assuming I have 2 node types, "paper" with 3 features, "price" with 1 feature, I have defined the feature matrix for both:

`features_paper = [{"node": 0, "name":"1","year":2000,"page":1},{"node": 1, "name":"2","year":2000,"page":1}]
x_paper = [None] * len(features_paper)
features_price= [{"node": 2, "value":10},{"node": 3, "value":10}]
x_price = [None] * len(features_price)

for feature in features_paper:
    x_paper[feature["node"]] = [int(feature["name"]), feature["year"], feature["page"]]
x_paper = torch.tensor(x_paper)
data['paper'].x =x_paper

for feature in features_price:
    x_price[feature["node"]-len(features_paper)] = [feature["value"]]
x_price = torch.tensor(x_price)
data['price'].x =x_price
print(data['paper'])`

I want to connect paper 0 (node 0) to paper 1 (node 1), and paper 0 to price 0 (node2), I define the edge_index like:

`data['paper', 'connect_to', 'paper'].edge_index=torch.tensor([[0,1],[1,0]],dtype=torch.long) # paper 0 (node 0) to paper 1 (node 1)
data['paper', 'connect_to', 'price'].edge_index=([[0,0],[0,0]],dtype=torch.long) # paper 0 (node 0) to price 0 (node 2)`

For same type nodes, e.g. "paper", if I write [0,1] in edge index, it means an edge is generated between the first and second tensor element in the feature matrix of "paper", i.e. node 0 and node 1. 
For different type nodes, e.g. "paper" to "price", if I write [0,0] in edge index, it means an edge is generated between the first element in the feature matrix of "paper" and the first element in the feature matrix of "price" i.e. node 0 and node 3.

Do I understand correctly?

**Reply by rusty1s** [reply]
Yes, your understanding is correct. To add reverse edges from `price` to `paper`, you would need to create a reverse edge type though:
```python
data['paper', 'connect_to', 'paper'].edge_index=torch.tensor([[0,1],[1,0]])
data['paper', 'connect_to', 'price'].edge_index=([[0,0]])
data['price', 'connect_to', 'paper'].edge_index=([[0,0]])
```
If you don't want to add reverse edges on your own, you can also leverage the `ToUndirected` transforms afterwards to add them later on.

**Reply by zzzLemon** [reply]
> Yes, your understanding is correct. To add reverse edges from `price` to `paper`, you would need to create a reverse edge type though:
> 
> ```python
> data['paper', 'connect_to', 'paper'].edge_index=torch.tensor([[0,1],[1,0]])
> data['paper', 'connect_to', 'price'].edge_index=([[0,0]])
> data['price', 'connect_to', 'paper'].edge_index=([[0,0]])
> ```
> 
> If you don't want to add reverse edges on your own, you can also leverage the `ToUndirected` transforms afterwards to add them later on.

Many thanks!
